macros/niceTables.pl

```diff
@@ -1,837 +1,987 @@
-				unless (ref($dataref->[$i][$j]) eq "HASH" or ref($dataref->[$i][$j]) eq "ARRAY");
-			# and if it was entered as an array reference, make the hash
-			if (ref($dataref->[$i][$j]) eq "ARRAY") {
-				my $temp = $dataref->[$i][$j];
-				$dataref->[$i][$j] = { "data", @$temp };
-			}
-			#before [a, options] was an option, {d=>a,options} was a shortcut for {data=>a,options}
-			${ $dataref->[$i][$j] }{data} = ${ $dataref->[$i][$j] }{d} if (defined ${ $dataref->[$i][$j] }{d});
-			# set default values for cell
-			${ $dataref->[$i][$j] }{header}    = ''         unless (defined ${ $dataref->[$i][$j] }{header});
-			${ $dataref->[$i][$j] }{tex}       = ''         unless (defined ${ $dataref->[$i][$j] }{tex});
-			${ $dataref->[$i][$j] }{b}         = 0          unless (defined ${ $dataref->[$i][$j] }{b});
-			${ $dataref->[$i][$j] }{i}         = 0          unless (defined ${ $dataref->[$i][$j] }{i});
-			${ $dataref->[$i][$j] }{m}         = 0          unless (defined ${ $dataref->[$i][$j] }{m});
-			${ $dataref->[$i][$j] }{noencase}  = 0          unless (defined ${ $dataref->[$i][$j] }{noencase});
-			${ $dataref->[$i][$j] }{halign}    = ''         unless (defined ${ $dataref->[$i][$j] }{halign});
-			${ $dataref->[$i][$j] }{colspan}   = ''         unless (defined ${ $dataref->[$i][$j] }{colspan});
-			${ $dataref->[$i][$j] }{cellcss}   = ''         unless (defined ${ $dataref->[$i][$j] }{cellcss});
-			${ $dataref->[$i][$j] }{rowcss}    = ''         unless (defined ${ $dataref->[$i][$j] }{rowcss});
-			${ $dataref->[$i][$j] }{rowcolor}  = ''         unless (defined ${ $dataref->[$i][$j] }{rowcolor});
-			${ $dataref->[$i][$j] }{midrule}   = 0          unless (defined ${ $dataref->[$i][$j] }{midrule});
-			${ $dataref->[$i][$j] }{headerrow} = 0          unless (defined ${ $dataref->[$i][$j] }{headerrow});
-			${ $dataref->[$i][$j] }{texencase} = [ '', '' ] unless (defined ${ $dataref->[$i][$j] }{texencase});
-			${ $dataref->[$i][$j] }{texpre}    = ${ $dataref->[$i][$j] }{texencase}->[0]
-				unless (defined ${ $dataref->[$i][$j] }{texpre});
-			${ $dataref->[$i][$j] }{texpost} = ${ $dataref->[$i][$j] }{texencase}->[1]
-				unless (defined ${ $dataref->[$i][$j] }{texpost});
+	my $userArray = shift;
+	my $dataArray = DataArray($userArray);
+	my $optsArray = OptionsArray($userArray);
+	my $colCount = ColumnCount($optsArray);
+	my $tableOpts = TableOptions($colCount, @_);
+	my $alignment = ParseAlignment($tableOpts->{texalignment});
+	TableEnvironment($dataArray, $optsArray, $colCount, $tableOpts, $alignment);
+}
+
+sub LayoutTable {
+	return DataTable(@_, LaYoUt => 1); 
+}
+
+# Make the outer table enovornment
+# Handle center, caption, horizontalrules, texalignment, Xratio
+# overall halign OK for tex, ptx but for html must be passed to cells
+# vertical rules from alingment OK
+# encase, rowheaders should be passed to cells
+# various css should be self-explanatory
+sub TableEnvironment {
+	my $dataArray = shift;
```

Comment:
We're trying to switch to:
```
my ($dataArray, $optsArray, $colCount, $tableOpts, $alignment) = @_; 
```
to pull arguments from the function calls.  It's shorter and this way this ensures there are no other arguments later in the function.